File: src/action_mgr.py
from enum import Enum

from tello_msgs.msg import TelloResponse
from tello_msgs.srv import TelloAction


class ActionMgr(object):
    """
    tello_driver uses a ROS service plus a ROS topic to simulate an action.
    This should be replaced when the rclpy action API appears in ROS2 Dashing.

    Overall flow:
        1. flock_base sends a TelloAction.Request on the tello_action service
        2. the drone responds with a TelloAction.Response
        3. later, the drone sends a TelloResponse message on the tello_response topic

    Only one action can be active at a time.

    TODO time out if tello_driver hasn't responded in ~10s, this can happen if tello_driver was restarted
    """

    class States(Enum):
        NOT_SENT = 0                    # TelloAction.Request not sent yet
        WAITING_FOR_FUTURE = 1          # Waiting for TelloAction.Response
        WAITING_FOR_RESPONSE = 2        # Waiting for TelloResponse
        SUCCEEDED = 3                   # Action succeeded, see the result string
        FAILED = 4                      # Action failed, see the result string
        FAILED_LOST_CONNECTION = 5      # Action failed, there's no connection to the drone

    def __init__(self, logger, action_client, action_code, action_str):
        self.logger = logger                        # Node.get_logger()
        self.action_client = action_client          # Node.create_client(...)
        self.action_code = action_code              # Action code, for reference
        self.action_str = action_str                # Action string, will be sent to the drone
        self.state = ActionMgr.States.NOT_SENT      # Internal state
        self.future = None                          # Future returned from client.call_async
        self.result_str = None                      # Result string

    def advance(self) -> States:
        """
        Call periodically to advance the state
        """
        if self.state == ActionMgr.States.NOT_SENT:
            self.logger.debug('send {} to tello_driver'.format(self.action_str))
            self.future = self.action_client.call_async(TelloAction.Request(cmd=self.action_str))
            self.state = ActionMgr.States.WAITING_FOR_FUTURE

        if self.state == ActionMgr.States.WAITING_FOR_FUTURE and self.future.done():
            response: TelloAction.Response = self.future.result()
            self.future = None

            if response.rc == TelloAction.Response.OK:
                self.logger.debug('{} accepted'.format(self.action_str))
                self.state = ActionMgr.States.WAITING_FOR_RESPONSE

            elif response.rc == TelloAction.Response.ERROR_BUSY:
                self.logger.error('{} failed, drone is busy'.format(self.action_str))
                self.result_str = 'drone is busy'
                self.state = ActionMgr.States.FAILED

            elif response.rc == TelloAction.Response.ERROR_NOT_CONNECTED:
                self.logger.error('{} failed, lost connection'.format(self.action_str))
                self.result_str = 'lost connection'
                self.state = ActionMgr.States.FAILED_LOST_CONNECTION

        return self.state

    def complete(self, msg: TelloResponse) -> States:
        """
        The tello_response message may arrive before self._future.done() evaluates True -- that's OK
        """
        if self.state != ActionMgr.States.WAITING_FOR_FUTURE and self.state != ActionMgr.States.WAITING_FOR_RESPONSE:
            self.logger.error('unexpected response {}'.format(msg))
            self.result_str = 'unexpected response'
            self.state = ActionMgr.States.FAILED

        elif msg.rc == TelloResponse.OK:
            self.logger.debug('{} succeeded with {}'.format(self.action_str, msg.str))
            self.result_str = msg.str
            self.state = ActionMgr.States.SUCCEEDED

        elif msg.rc == TelloResponse.ERROR:
            self.logger.error('{} failed with {}'.format(self.action_str, msg.str))
            self.result_str = msg.str
            self.state = ActionMgr.States.FAILED

        elif msg.rc == TelloResponse.TIMEOUT:
            self.logger.error('{} failed, drone timed out'.format(self.action_str))
            self.result_str = 'tello_driver timed out'
            self.state = ActionMgr.States.FAILED

        return self.state

File: src/action_mgr.cpp
#include "action_mgr.hpp"

// TODO time out if tello_driver hasn't responded in ~10s, this can happen if tello_driver was restarted

namespace drone_base {

ActionMgr::State ActionMgr::send(Action action, std::string action_str)
{
  action_ = action;
  action_str_ = action_str;

  RCLCPP_DEBUG(logger_, "send %s to tello_driver", action_str.c_str());

  auto request = std::make_shared<tello_msgs::srv::TelloAction::Request>();
  request->cmd = action_str;
  future_ = client_->async_send_request(request);

  state_ = ActionMgr::State::waiting_for_future;
  return state_;
}

ActionMgr::State ActionMgr::spin_once()
{
  if (state_ == ActionMgr::State::waiting_for_future &&
    future_.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {

    // Get the initial response
    tello_msgs::srv::TelloAction::Response::SharedPtr response = future_.get();

    if (response->rc == response->OK) {
      RCLCPP_DEBUG(logger_, "%s accepted", action_str_.c_str());
      state_ = ActionMgr::State::waiting_for_response;

    } else if (response->rc == response->ERROR_BUSY) {
      RCLCPP_ERROR(logger_, "%s failed, drone is busy", action_str_.c_str());
      result_str_ = "drone is busy";
      state_ = ActionMgr::State::failed;

    } else if (response->rc == response->ERROR_NOT_CONNECTED) {
      RCLCPP_ERROR(logger_, "%s failed, lost connection", action_str_.c_str());
      result_str_ = "lost connection";
      state_ = ActionMgr::State::failed_lost_connection;
    }
  }

  return state_;
}

ActionMgr::State ActionMgr::complete(tello_msgs::msg::TelloResponse::SharedPtr msg)
{
  // The tello_response message may arrive before the future is ready -- that's OK
  if (!busy()) {
    RCLCPP_ERROR(logger_, "unexpected response %s", msg->str.c_str());
    result_str_ = "unexpected response";
    state_ = ActionMgr::State::failed;

  } else if (msg->rc == msg->OK) {
    RCLCPP_DEBUG(logger_, "%s succeeded with %s", action_str_.c_str(), msg->str.c_str());
    result_str_ = msg->str;
    state_ = ActionMgr::State::succeeded;

  } else if (msg->rc == msg->ERROR) {
    RCLCPP_ERROR(logger_, "%s failed with %s", action_str_.c_str(), msg->str.c_str());
    result_str_ = msg->str;
    state_ = ActionMgr::State::failed;

  } else if (msg->rc == msg->TIMEOUT) {
    RCLCPP_ERROR(logger_, "%s failed, drone timed out", msg->str.c_str());
    result_str_ = "drone timed out";
    state_ = ActionMgr::State::failed;
  }

  return state_;
}

} // namespace drone_base